References
----------

.. [Trisicell] Mehrabadi *et al.* (2021),
   *Trisicell: Scalable Tumor Phylogeny Reconstruction and Validation Reveals Developmental Origin and Therapeutic Impact of Intratumoral Heterogeneity*

.. [ReviewBinary] Malikic *et al.* (2020),
   *Studying the history of tumor evolution from single-cell sequencing data by exploring the space of binary matrices*,
   `bioRxiv <https://doi.org/10.1101/2020.07.15.204081>`__.

.. [PhISCS-BnB] Sadeqi Azer *et al.* (2020),
   *PhISCS-BnB: a fast branch and bound algorithm for the perfect tumor phylogeny reconstruction problem*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btaa464>`__.

.. [PhISCS] Malikic *et al.* (2019),
   *PhISCS: a combinatorial approach for subperfect tumor phylogeny reconstruction via integrative use of single-cell and bulk sequencing data*,
   `Genome Research <http://doi.org/10.1101/gr.234435.118>`__.

.. [ScisTree] Wu (2019),
   *Accurate and efficient cell lineage tree inference from noisy single cell data: the maximum likelihood perfect phylogeny approach*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btz676>`__.

.. [SCITE] Jahn *et al.* (2016),
   *Tree inference for single-cell data*,
   `Genome Biology <https://doi.org/10.1186/s13059-016-0936-x>`__.

.. [OncoNEM] Ross *et al.* (2016),
   *OncoNEM: inferring tumor evolution from single-cell sequencing data*,
   `Genome Biology <https://doi.org/10.1186/s13059-016-0929-9>`__.